// @open-pets/client/dist/index.js
import net from "node:net";
import { randomUUID } from "node:crypto";
import { parseIpcEndpoint, readDiscoveryFile } from "./discovery.js";
import { connectTimeoutMs, maxIpcMessageBytes, openPetsIpcVersion, parseIpcResponse, responseTimeoutMs, validateReaction, OpenPetsClientError } from "./protocol.js";
export { getDiscoveryFilePath, parseIpcEndpoint, readDiscoveryFile, validateDiscovery, validateEndpoint } from "./discovery.js";
export { allowedReactions, OpenPetsClientError } from "./protocol.js";
export function createOpenPetsClient(options = {}) {
    return {
        hello: () => sendDiscoveredRequest("hello", {}, options),
        status: async (statusOptions) => {
            try {
                return await sendDiscoveredRequest("status", { leaseId: statusOptions?.leaseId }, options);
            }
            catch (error) {
                return {
                    ok: false,
                    appRunning: false,
                    unavailableReason: error instanceof Error ? error.message : "OpenPets is unavailable.",
                };
            }
        },
        listPets: async () => parsePetListResult(await sendDiscoveredRequest("pets.list", {}, options)),
        installPet: async (petId) => parsePetInstallResult(await sendDiscoveredRequest("pets.install", { petId: validatePetId(petId) }, { ...options, responseTimeoutMs: options.responseTimeoutMs ?? 60_000 })),
        acquireLease: (leaseOptions) => sendDiscoveredRequest("lease.acquire", { requestedPetId: leaseOptions?.requestedPetId }, options),
        heartbeatLease: (leaseId) => sendDiscoveredRequest("lease.heartbeat", { leaseId }, options),
        releaseLease: (leaseId) => sendDiscoveredRequest("lease.release", { leaseId }, options),
        react: (reaction, reactOptions) => sendDiscoveredRequest("pet.react", { reaction: validateReaction(reaction), leaseId: reactOptions?.leaseId }, options),
        say: (message, sayOptions) => sendDiscoveredRequest("pet.say", { message, reaction: sayOptions?.reaction, leaseId: sayOptions?.leaseId }, options),
    };
}
export function parsePetInstallResult(value) {
    if (!isRecord(value) || value.ok !== true || typeof value.petId !== "string" || typeof value.displayName !== "string" || value.installed !== true) {
        throw new OpenPetsClientError("invalid_response", "OpenPets pet install response is invalid.");
    }
    return { ok: true, petId: value.petId, displayName: value.displayName, installed: true };
}
function validatePetId(value) {
    if (!/^[a-z0-9][a-z0-9_-]{0,63}$/.test(value) || value === "builtin") {
        throw new OpenPetsClientError("invalid_pet_id", "Invalid OpenPets pet id.");
    }
    return value;
}
export function parsePetListResult(value) {
    if (!isRecord(value) || value.ok !== true || !Array.isArray(value.pets) || typeof value.defaultPetId !== "string") {
        throw new OpenPetsClientError("invalid_response", "OpenPets pet list response is invalid.");
    }
    return {
        ok: true,
        defaultPetId: value.defaultPetId,
        pets: value.pets.map(parsePetListItem),
    };
}
function parsePetListItem(value) {
    if (!isRecord(value) || typeof value.id !== "string" || typeof value.displayName !== "string" || typeof value.builtIn !== "boolean" || typeof value.broken !== "boolean") {
        throw new OpenPetsClientError("invalid_response", "OpenPets pet list item is invalid.");
    }
    return { id: value.id, displayName: value.displayName, builtIn: value.builtIn, broken: value.broken };
}
function isRecord(value) {
    return typeof value === "object" && value !== null;
}
async function sendDiscoveredRequest(method, params, options) {
    const discovery = readDiscoveryFile(options.discoveryPath);
    return sendRequest(discovery, method, params, options);
}
export function sendRequest(discovery, method, params, options = {}) {
    const request = {
        id: randomUUID(),
        version: openPetsIpcVersion,
        token: discovery.token,
        method,
        params,
    };
    const requestLine = `${JSON.stringify(request)}\n`;
    if (Buffer.byteLength(requestLine, "utf8") > maxIpcMessageBytes) {
        return Promise.reject(new OpenPetsClientError("request_too_large", "OpenPets IPC request is too large."));
    }
    return new Promise((resolve, reject) => {
        const endpoint = parseIpcEndpoint(discovery.endpoint);
        const socket = endpoint.kind === "tcp" ? net.createConnection({ host: endpoint.host, port: endpoint.port }) : net.createConnection(endpoint.path);
        let buffer = "";
        let settled = false;
        const connectTimer = setTimeout(() => finish(new OpenPetsClientError("connect_timeout", "Timed out connecting to OpenPets.")), options.connectTimeoutMs ?? connectTimeoutMs);
        const responseTimer = setTimeout(() => finish(new OpenPetsClientError("response_timeout", "Timed out waiting for OpenPets response.")), options.responseTimeoutMs ?? responseTimeoutMs);
        const finish = (error, result) => {
            if (settled)
                return;
            settled = true;
            clearTimeout(connectTimer);
            clearTimeout(responseTimer);
            socket.destroy();
            if (error)
                reject(error);
            else
                resolve(result);
        };
        socket.setEncoding("utf8");
        socket.once("connect", () => {
            clearTimeout(connectTimer);
            socket.write(requestLine);
        });
        socket.on("data", (chunk) => {
            buffer += chunk;
            if (Buffer.byteLength(buffer, "utf8") > maxIpcMessageBytes) {
                finish(new OpenPetsClientError("response_too_large", "OpenPets IPC response is too large."));
                return;
            }
            const newline = buffer.indexOf("\n");
            if (newline === -1)
                return;
            try {
                const parsed = parseIpcResponse(JSON.parse(buffer.slice(0, newline)));
                if (parsed.ok)
                    finish(undefined, parsed.result);
                else
                    finish(new OpenPetsClientError(parsed.error.code, parsed.error.message));
            }
            catch (error) {
                finish(error);
            }
        });
        socket.once("error", (error) => finish(new OpenPetsClientError("unavailable", error.message)));
        socket.once("end", () => {
            if (!settled)
                finish(new OpenPetsClientError("connection_closed", "OpenPets closed the IPC connection before responding."));
        });
    });
}
