// @open-pets/client/dist/protocol.js
export const openPetsIpcProtocol = "openpets-ipc";
export const openPetsIpcVersion = 1;
export const maxIpcMessageBytes = 16 * 1024;
export const connectTimeoutMs = 2_000;
export const responseTimeoutMs = 3_000;
export const allowedReactions = [
    "idle",
    "thinking",
    "working",
    "editing",
    "running",
    "testing",
    "waiting",
    "waving",
    "success",
    "error",
    "celebrating",
];
export function parseIpcResponse(value) {
    if (!isRecord(value))
        throw new OpenPetsClientError("invalid_response", "IPC response must be an object.");
    if (typeof value.id !== "string" && value.id !== null)
        throw new OpenPetsClientError("invalid_response", "IPC response id is invalid.");
    if (value.ok === true) {
        return { id: value.id, ok: true, result: value.result };
    }
    if (value.ok === false && isRecord(value.error) && typeof value.error.code === "string" && typeof value.error.message === "string") {
        return { id: value.id, ok: false, error: { code: value.error.code, message: value.error.message } };
    }
    throw new OpenPetsClientError("invalid_response", "IPC response shape is invalid.");
}
export function validateReaction(value) {
    if (!allowedReactions.includes(value)) {
        throw new OpenPetsClientError("invalid_reaction", "Invalid OpenPets reaction.");
    }
    return value;
}
export class OpenPetsClientError extends Error {
    code;
    constructor(code, message) {
        super(message);
        this.code = code;
    }
}
export function isRecord(value) {
    return typeof value === "object" && value !== null;
}
